File: sdk/go/modules/dht/dht.go
// Package dht provides a typed wrapper for the CONDUYT DHT11/DHT22 module.
package dht

import (
	"context"
	"encoding/binary"
	"fmt"
	"math"

	conduyt "github.com/virgilvox/conduyt/sdk/go"
)

// Reading is one DHT temperature/humidity sample.
type Reading struct {
	TemperatureC float32
	HumidityPct  float32
}

// DHT provides typed access to a DHT11/DHT22 sensor.
type DHT struct {
	device   *conduyt.Device
	moduleID byte
}

// New creates a DHT wrapper for the given device and module ID.
func New(device *conduyt.Device, moduleID byte) *DHT {
	return &DHT{device: device, moduleID: moduleID}
}

// Begin initializes the sensor. typ is 11 for DHT11 or 22 for DHT22.
func (d *DHT) Begin(ctx context.Context, pin, typ byte) error {
	_, err := d.device.ModCmd(ctx, []byte{d.moduleID, 0x01, pin, typ})
	return err
}

// Read triggers a sample and returns temperature + humidity. The first byte
// of the MOD_RESP payload is the module id; the firmware emits two LE
// float32s after that.
func (d *DHT) Read(ctx context.Context) (Reading, error) {
	resp, err := d.device.ModCmd(ctx, []byte{d.moduleID, 0x02})
	if err != nil {
		return Reading{}, err
	}
	if len(resp) < 9 {
		return Reading{}, fmt.Errorf("dht: short response (%d bytes)", len(resp))
	}
	t := math.Float32frombits(binary.LittleEndian.Uint32(resp[1:5]))
	h := math.Float32frombits(binary.LittleEndian.Uint32(resp[5:9]))
	return Reading{TemperatureC: t, HumidityPct: h}, nil
}

File: sdk/rust/src/modules/dht.rs
//! CONDUYT DHT11/DHT22 Module — typed wrapper

use crate::device::{Device, DeviceError};
use crate::transports::Transport;

/// One DHT temperature/humidity sample.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Reading {
    pub temperature_c: f32,
    pub humidity_pct: f32,
}

/// DHT module wrapper.
pub struct DHT<'a, T: Transport> {
    device: &'a mut Device<T>,
    module_id: u8,
}

impl<'a, T: Transport> DHT<'a, T> {
    pub fn new(device: &'a mut Device<T>, module_id: u8) -> Self {
        Self { device, module_id }
    }

    /// Initialize the sensor. `kind = 11` for DHT11, `22` for DHT22.
    pub fn begin(&mut self, pin: u8, kind: u8) -> Result<(), DeviceError<T::Error>> {
        self.device.mod_cmd(&[self.module_id, 0x01, pin, kind])?;
        Ok(())
    }

    /// Trigger a sample and decode the temperature + humidity floats. The
    /// MOD_RESP echoes the module id as byte 0, then two LE float32s.
    pub fn read(&mut self) -> Result<Reading, DeviceError<T::Error>> {
        let resp = self.device.mod_cmd(&[self.module_id, 0x02])?;
        if resp.len() < 9 {
            // Treat short response as a malformed wire frame.
            return Err(DeviceError::WireError(crate::wire::WireError::IncompletePkt));
        }
        let t = f32::from_le_bytes([resp[1], resp[2], resp[3], resp[4]]);
        let h = f32::from_le_bytes([resp[5], resp[6], resp[7], resp[8]]);
        Ok(Reading { temperature_c: t, humidity_pct: h })
    }
}
